### PR TITLE
Caroline Nardi - Ride Share - Octos

### PR DESCRIPTION
**Pull request by cmn53**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | One design decision that I made was deciding how much functionality should be included within the TripDispatcher#request_trip method, particularly in Wave 2 when the next driver was just the first available driver in the CSV. Including the code for the new_trip_id and select_drivers within the request_method seemed like a viable option since these were both fairly short calculations, but I decided it made more sense to break them out into separate methods, particularly since the select_driver method would be altered significantly in Wave 3.
Describe any examples of composition that you encountered in this project, if any | TripDispatcher has collections of trips, drivers, and passengers; trips have passengers and drivers; and passengers and drivers have collections of trips.
Describe a _nominal test_ that you wrote for this assignment. | I wrote a nominal test to make sure that total revenue would account for all of a driver's completed trips. 
Describe an _edge case test_ that you wrote for this assignment. | I wrote an edge case test to make sure that a driver could not have negative revenue if the trip cost was less than the trip fee.
Describe a concept that you gained more clarity on as you worked on this assignment. | I think I gained a little more clarity on how objects could be initiated in large batches and how a single 'admin' class could be used to navigate the attributes of other related and interdependent classes.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -21,27 +20,68 @@ def initialize(input)
...
+      seconds_per_hour = 3600.0
+      total_revenue_hrs = total_time / seconds_per_hour
+      revenue_per_hour = total_revenue / total_revenue_hrs
+
+      return revenue_per_hour
+    end
+
+    def completed_trips
+      @trips.select { |trip| trip.is_finished? }
     end
```
Comment (droberts-sea) [inline lib/driver.rb:85]: I love that you've broken this out as a separate helper method! This makes the other code in this file much easier to read.

File: lib/trip_dispatcher.rb [hunk 2]
```diff
@@ -87,7 +87,72 @@ def load_trips
...
+    end
+
+    def new_trip_id
+      @trips.map { |trip| trip.id }.max + 1
+    end
+
+    def available_drivers
+      available_drivers = @drivers.select do |driver|
+        driver.status == :AVAILABLE && driver.trips.length > 0
+      end
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:127]: Again, great use of helper methods in this file - these feel like they're just the right size and level of complexity.

File: specs/driver_spec.rb [hunk 3]
```diff
@@ -37,44 +37,183 @@
...
+      @driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV", vin: "1C9EVBRM0YBC564DZ")
+    end
+
+    it "returns a number" do
+      @driver.total_time.must_be_kind_of Numeric
+    end
+
+    it "returns zero if driver has no completed trips" do
+      @driver.total_time.must_equal 0
+
```
Comment (droberts-sea) [inline specs/driver_spec.rb:136]: For these wave 1 methods, I would also be interested in what happens if the passenger or driver has an incomplete trip.

File: specs/trip_dispatch_spec.rb [hunk 4]
```diff
@@ -89,4 +89,146 @@
...
+      dispatcher.available_drivers.each do |driver|
+        driver.must_be_instance_of RideShare::Driver
+        driver.status.must_equal :AVAILABLE
+      end
+    end
+
+    it "raises an error if all drivers are unavailable" do
+      dispatcher = RideShare::TripDispatcher.new
+      dispatcher.drivers.each do |driver|
+        if driver.status == :AVAILABLE
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:162]: Good work getting this tricky edge case figured out.

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | yes
**Wave 1** |
`Trip` has a helper method to calculate duration | yes
`Passenger` has a method to calculate total cost of all trips | yes
`Driver` has a method to calculate total revenue | yes
Complete tests | yes
**Wave 2** |
`TripDispatcher` has a new method to create trips | yes
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | yes
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | yes
Complete tests | yes
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | yes
Appropriate use of Ruby's `Date` | yes
Appropriate helper methods were made to help with complex logic | yes
Complete tests | yes
**Overall** | Excellent job overall! Your code is clear and easy to understand, your logic is sound, and your test coverage is solid. I have left a few nitpicks below, but in general I am quite pleased with this submission, and excited to see what you come up with on Hotel. Keep up the hard work!
